### PR TITLE
cpu/stm32: unified ADC driver implementations

### PR DESCRIPTION
**Pull request by haukepetersen**
EDIT
~~This PR is still work in progress, just wanted to put it already out there to prevent multiple people working on this...~~

I have not test all of the STM CPUs, yet. ~~Especially I doubt the L1 and L4's to work, as it seems they have subtle differences in their ADC registers...~~

L1 works, L4 was not implemented before anyway...

### CODE REVIEW

**Comment by aabadie** [comment]
> just wanted to put it already out there to prevent multiple people working on this...

good idea, I was thinking of this some times ago :)

**Comment by aabadie** [comment]
this already needs rebase btw...

**Comment by haukepetersen** [comment]
rebased

File: boards/msbiot/include/periph_conf.h [hunk 1]
```diff
@@ -96,17 +96,14 @@ static const pwm_conf_t pwm_config[] = {
...
- * We need to define the following fields:
- * PIN, device (ADCx), channel
  * @{
  */
-#define ADC_CONFIG {             \
-    {GPIO_PIN(PORT_B, 0), 0, 8}, \
-    {GPIO_PIN(PORT_B, 1), 0, 9}  \
-}
+static const adc_conf_t adc_config[] = {
+    { .pin = GPIO_PIN(PORT_B, 0), .dev = ADC_1, .chan =  8 },
```
Comment (aabadie) [inline boards/msbiot/include/periph_conf.h]: extra space before 8 (and 9 the line below)

File: boards/nucleo32-common/include/periph_conf_nucleo32.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ */
+static const adc_conf_t adc_config[] = {
+    { .pin = GPIO_PIN(PORT_A, 0), .dev = ADC_1, .chan = 0 },
+    { .pin = GPIO_PIN(PORT_A, 1), .dev = ADC_1, .chan = 1 },
+    { .pin = GPIO_PIN(PORT_A, 3), .dev = ADC_1, .chan = 3 },
+    { .pin = GPIO_PIN(PORT_A, 4), .dev = ADC_1, .chan = 4 },
+    { .pin = GPIO_PIN(PORT_A, 5), .dev = ADC_1, .chan = 5 },
+    { .pin = GPIO_PIN(PORT_A, 6), .dev = ADC_1, .chan = 6 },
+    { .pin = GPIO_PIN(PORT_A, 7), .dev = ADC_1, .chan = 7 },
+    { .pin = GPIO_PIN(PORT_A, 2), .dev = ADC_1, .chan = 2 }
```
Comment (aabadie) [inline boards/nucleo32-common/include/periph_conf_nucleo32.h]: Maybe move this line at third position

File: boards/nucleo32-f031/include/periph_conf.h [hunk 3]
```diff
@@ -174,21 +174,6 @@ static const spi_conf_t spi_config[] = {
 #define RTC_NUMOF           (0U)
 /** @} */
 
-/**
- * @name   ADC configuration
- * @{
- */
-#define ADC_CONFIG {            \
```
Comment (aabadie) [inline boards/nucleo32-f031/include/periph_conf.h]: This seems very different from the common part.

File: boards/nucleo32-common/include/periph_conf_nucleo32.h (hunk 2 above)
Comment (haukepetersen) [inline boards/nucleo32-common/include/periph_conf_nucleo32.h]: nope, the way I put them here is so that the Ardunio Ax lines match RIOTs ADC_LINE(x) configuration. ST simply mapped their ADCs like this...

File: boards/msbiot/include/periph_conf.h (hunk 1 above)
Comment (haukepetersen) [inline boards/msbiot/include/periph_conf.h]: will fix

File: boards/nucleo32-f031/include/periph_conf.h (hunk 3 above)
Comment (haukepetersen) [inline boards/nucleo32-f031/include/periph_conf.h]: I don't understand, what do you mean with common part?

**Comment by haukepetersen** [comment]
> I would prefer to provide the second point in dedicated PRs (even if I like the general idea).

I don't like to make this a separate PR -> IMHO it is good practice to start the config centralization in a concrete PR (as in this one), so we have already some 'meat' to put in the common files and it is clear how the centralization is intended to work...

File: boards/nucleo32-f031/include/periph_conf.h (hunk 3 above)
Comment (smlng) [inline boards/nucleo32-f031/include/periph_conf.h]: I guess @aabadie means, currently there are 5 ADC lines defined here, but in your common definition in `periph_conf_nucleo32.h` above you define 8 ADC lines.

File: boards/opencm904/include/periph_conf.h [hunk 4]
```diff
@@ -60,13 +60,6 @@ extern "C" {
 /** @} */
 
 /**
- * @name ADC configuration
- * @{
- */
-#define ADC_NUMOF           (0)
```
Comment (smlng) [inline boards/opencm904/include/periph_conf.h]: unrelated?

File: boards/spark-core/include/periph_conf.h [hunk 5]
```diff
@@ -60,13 +60,6 @@
 /** @} */
 
 /**
- * @name ADC configuration
- * @{
- */
-#define ADC_NUMOF           (0)
```
Comment (smlng) [inline boards/spark-core/include/periph_conf.h]: dito, seems unrelated

File: boards/opencm904/include/periph_conf.h (hunk 4 above)
Comment (haukepetersen) [inline boards/opencm904/include/periph_conf.h]: nope, related: the opencm904 uses a stm cpu and the reworked adc driver does not requite 'empty' numof defs anymore... So related cleanup.

File: boards/spark-core/include/periph_conf.h (hunk 5 above)
Comment (haukepetersen) [inline boards/spark-core/include/periph_conf.h]: same here.

File: boards/nucleo32-f031/include/periph_conf.h (hunk 3 above)
Comment (haukepetersen) [inline boards/nucleo32-f031/include/periph_conf.h]: yes, but as I see it, the existing config was incomplete and this PR fixes that...

Comment (smlng) [inline boards/nucleo32-f031/include/periph_conf.h]: might be related to the fact that SPI1 and some PWM channels are on some of the ADC lines, too. But even the original ADC config does not ignore those pins and would hence collide with those, too.

**Comment by smlng** [comment]
@haukepetersen what's your ETA on this one? Further, I think L1-L4 should be adaptable to this, too - maybe a bit more ifdef, but as there is already that much ...

**Comment by haukepetersen** [comment]
ETA: as soon as possible, but not before beginning of August, as there is the Dagstuhl workshop next week...

**Comment by haukepetersen** [comment]
So, finally got to this PR again - sorry for the delay.

The PR should now be ready for review/testing. All boards that had the ADC feature before should still work. Just tested with `nucleo-f072`, `nucleo-f411`, and `nucleo-l152`.

ADC drivers are missing for `stm32l0`, `stm32l4`, and `stm32f7`, but these never existed so far and merging this into the unified driver is work for another day...

**Comment by haukepetersen** [comment]
Picked up work on this again, update will follow soon

File: cpu/stm32_common/periph/adc.c [hunk 6]
```diff
@@ -1,67 +1,77 @@
...
+#define DEVOFFSET               (8)
+
+#if defined(CPU_FAM_STM32L1)
+#define WRITE_LINE(x)           (dev(line)->SQR5 = x)
+#else
+#define WRITE_LINE(x)           (dev(line)->SQR3 = x)
+#endif
+
+/* find out which bus the ADC is connected to */
+#if defined()
```
Comment (aabadie) [inline cpu/stm32_common/periph/adc.c]: Looks like there are things missing here

**Comment by haukepetersen** [comment]
yes, its still work in progress. Run into trouble with some STM CPUs and never found the time to continue debugging... But I havn't forgotten about this PR :-)

**Comment by aabadie** [comment]
#7815 has been merged so maybe this PR could be adapted with it ?

**Comment by haukepetersen** [comment]
Just saw that... So far I simply skipped support for the L0 (same as for the F0 and F3), so merging these I would suggest to be done in follow up PRs. Otherwise this PR gets out of hand.

So for now, this PR should be ready (except some restructuring of commits...). So please go ahead and re-test, as I did not test all the boards that are effected, yet.

**Comment by aabadie** [comment]
Gave this one a try:
* works on: nucleo-f030, nucleo-f446
* fails on:
  * nucleo-f334 (build fails on f3)
  * nucleo144-f207/nucleo144-f446: A3, A4 and A5 don't work. I guess the configuration is wrong for nucleo-144 boards

**Comment by haukepetersen** [comment]
thanks. The f3 won't work, as there is no ADC support for the F3, yet..
Need to look into whats going wrong for the f207 and the f446 boards...

**Comment by haukepetersen** [comment]
The nucleo144-f207 should not even build - no ADC feature defined...

**Comment by haukepetersen** [comment]
same for the `nucleo144-f446`...

**Comment by haukepetersen** [comment]
nope, everything correct there, simply miss- (or better non-) configured features for those two boards.

File: boards/nucleo144-f207/include/periph_conf.h [hunk 7]
```diff
@@ -258,16 +258,18 @@ static const spi_conf_t spi_config[] = {
...
- * We need to define the following fields:
- * PIN, device (ADCx), channel
  * @{
  */
-#define ADC_CONFIG {              \
-    {GPIO_PIN(PORT_A, 3), 0, 3},  \
-    {GPIO_PIN(PORT_C, 0), 1, 0}  \
-}
-#define ADC_NUMOF          (2)
+static const adc_conf_t adc_config[] = {
```
Comment (aabadie) [inline boards/nucleo144-f207/include/periph_conf.h]: This configuration is the same for nucleo144-f412 and nucleo144-f413, maybe they could be put in a shared common file ? in `boards/nucleo144-common/include/periph-common144.h` or something ?

nucleo144-f429 and nucleo144-f446 have a different one.

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 8]
```diff
@@ -206,6 +235,37 @@ typedef enum {
 #endif /* ndef CPU_FAM_STM32F1 */
 
 /**
+ * @brief   Override the ADC resolution configuration
+ * @{
```
Comment (aabadie) [inline cpu/stm32_common/include/periph_cpu_common.h]: This is not needed and the closing block below as well

Comment (aabadie) [reply cpu/stm32_common/include/periph_cpu_common.h]: Or the `@brief` should become `@name`

File: boards/nucleo32-common/include/periph_conf_nucleo32.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C)  2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     boards_nucleo-common
```
Comment (aabadie) [inline boards/nucleo32-common/include/periph_conf_nucleo32.h]: shouldn't it be boards_nucleo32-common ? (not sure)

**Comment by haukepetersen** [comment]
the configurations for both boards were wrong, also - hopefully fixed now.

File: boards/nucleo32-common/include/periph_conf_nucleo32.h (hunk 9 above)
Comment (aabadie) [reply boards/nucleo32-common/include/periph_conf_nucleo32.h]: yes it should (just checked)

File: boards/nucleo144-f207/include/periph_conf.h (hunk 7 above)
Comment (haukepetersen) [reply boards/nucleo144-f207/include/periph_conf.h]: we could do that, but I think if it is only those two, it makes sense to keep them as is. Otherwise we start with many files that are only applicable to random sub-groups of boards?!

File: cpu/stm32_common/include/periph_cpu_common.h (hunk 8 above)
Comment (haukepetersen) [reply cpu/stm32_common/include/periph_cpu_common.h]: true

File: boards/msbiot/include/periph_conf.h (hunk 1 above)
Comment (aabadie) [reply boards/msbiot/include/periph_conf.h]: there is still one

File: boards/nucleo32-common/include/periph_conf_nucleo32.h (hunk 9 above)
Comment (haukepetersen) [reply boards/nucleo32-common/include/periph_conf_nucleo32.h]: fixed

File: boards/nucleo144-f207/include/periph_conf.h (hunk 7 above)
Comment (aabadie) [reply boards/nucleo144-f207/include/periph_conf.h]: agreed, let's keep it like this

Comment (haukepetersen) [reply boards/nucleo144-f207/include/periph_conf.h]: perfect.

**Comment by aabadie** [comment]
Almost there:
* needs rebase again since #7964 has just been merged
* [one comment](https://github.com/RIOT-OS/RIOT/pull/7277/files#r124809635) still needs to be addressed
* I'll give it another try tomorrow

**Comment by haukepetersen** [comment]
- addressed comment
- rebased

File: boards/nucleo144-f207/include/periph_conf.h [hunk 10]
```diff
@@ -258,16 +258,18 @@ static const spi_conf_t spi_config[] = {
...
+    { .pin = GPIO_PIN(PORT_A,  3), .dev = ADC_1, .chan =  3 },
+    { .pin = GPIO_PIN(PORT_C,  0), .dev = ADC_1, .chan = 10 },
+    { .pin = GPIO_PIN(PORT_C,  3), .dev = ADC_1, .chan = 13 },
+    { .pin = GPIO_PIN(PORT_F,  3), .dev = ADC_3, .chan =  9 },
```
Comment (aabadie) [inline boards/nucleo144-f207/include/periph_conf.h]: The build fails for me because of the use of ADC_3. Even if it's compliant with the board datasheet.

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 11]
```diff
@@ -136,6 +150,21 @@ typedef uint32_t gpio_t;
...
+ */
+enum {
+    ADC_1 = 0,          /**< ADC1 */
+#if defined(CPU_FAM_STM32F1) || defined(CPU_FAM_STM32F2)
+    ADC_2 = 1,          /**< ADC2 */
+#endif
+#if defined(CPU_MODEL_STM32F407VG) || defined(CPU_MODEL_STM32F415RG) \
+    || defined(CPU_MODEL_STM32F446RE) || defined(CPU_MODEL_STM32F429ZI)
+    ADC_2 = 1,          /**< ADC2 */
+    ADC_3 = 2           /**< ADC3 */
```
Comment (aabadie) [inline cpu/stm32_common/include/periph_cpu_common.h]: ADC_3 is not defined for CPU_FAM_STM32F2 but is used in nucleo144-f207: the build fails. I tried to enable it for f207 but the application hangs when initializing ADC3...

File: boards/nucleo144-f429/include/periph_conf.h [hunk 12]
```diff
@@ -233,24 +233,19 @@ static const spi_conf_t spi_config[] = {
...
-    {GPIO_PIN(PORT_C, 3), 2, 13}, \
-    {GPIO_PIN(PORT_F, 3), 2, 9},  \
-    {GPIO_PIN(PORT_F, 5), 2, 15}, \
-    {GPIO_PIN(PORT_F, 10), 2, 8}, \
-}
+static const adc_conf_t adc_config[] = {
+    { .pin = GPIO_PIN(PORT_A,  3), .dev = ADC_1, .chan =  3 },
+    { .pin = GPIO_PIN(PORT_C,  0), .dev = ADC_1, .chan = 10 },
+    { .pin = GPIO_PIN(PORT_C,  3), .dev = ADC_1, .chan = 13 },
+    { .pin = GPIO_PIN(PORT_F,  3), .dev = ADC_1, .chan =  9 },
```
Comment (aabadie) [inline boards/nucleo144-f429/include/periph_conf.h]: According to the board datasheet, the device should be `ADC_3`. I tried to change that but then the application hangs when initializing it (similar issue with f207).

**Comment by haukepetersen** [comment]
Hmm, seems like I did not compile-test my last fix. I don't have any of these boards here, so I couldnt test on hardware...

Will take a look again at the ADC3 situation, shouldn't be too hard to find :-)

**Comment by fjmolinas** [comment]
@aabadie do you think this one is still relevant with the current stm32 structure?

**Comment by aabadie** [comment]
Having everything in a single `adc.c` file is not the way to go IMHO, we should take inspiration of what was done for i2c. The shared ADC config for some nucleo64 boards is something that could be kept.
In the current state, this PR is too much out-of-date, so better split-out parts of it in new PRs.

**Comment by haukepetersen** [comment]
I will not have the time to work on this any further... Please feel free to close/fork/take this PR or whatever might be useful of this :-)

**Comment by Teufelchen1** [comment]
Closed as out-of-date.
